synapse/storage/data_stores/main/ui_auth.py

```diff
@@ -0,0 +1,294 @@
+        # try a few times till one goes through, giving up eventually.
+        attempts = 0
+        while attempts < 5:
+            session_id = stringutils.random_string(24)
+
+            try:
+                await self.db.simple_insert(
+                    table="ui_auth_sessions",
+                    values={
+                        "session_id": session_id,
+                        "clientdict": clientdict_json,
+                        "uri": uri,
+                        "method": method,
+                        "description": description,
+                        "serverdict": "{}",
+                        "creation_time": self.hs.get_clock().time_msec(),
+                    },
+                    desc="create_ui_auth_session",
+                )
+                return UIAuthSessionData(
+                    session_id, clientdict, uri, method, description
+                )
+            except self.db.engine.module.IntegrityError:
+                attempts += 1
+        raise StoreError(500, "Couldn't generate a session ID.")
+
+    async def get_ui_auth_session(self, session_id: str) -> UIAuthSessionData:
+        """Retrieve a UI auth session.
+
+        Args:
+            session_id: The ID of the session.
+        Returns:
+            A dict containing the device information.
+        Raises:
+            SynapseError: if the session is not found.
+        """
+        try:
+            result = await self.db.simple_select_one(
+                table="ui_auth_sessions",
+                keyvalues={"session_id": session_id},
+                retcols=("clientdict", "uri", "method", "description"),
+                desc="get_ui_auth_session",
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: %s" % session_id)
+
+        result["clientdict"] = json.loads(result["clientdict"])
+
+        return UIAuthSessionData(session_id, **result)
+
+    async def mark_ui_auth_stage_complete(
+        self, session_id: str, stage_type: str, result: Union[str, bool, JsonDict],
+    ):
+        """
+        Mark a session stage as completed.
+
+        Args:
+            session_id: The ID of the corresponding session.
+            stage_type: The completed stage type.
+            result: The result of the stage verification.
+        """
+        await self.db.runInteraction(
+            "mark_ui_auth_stage_complete",
+            self._mark_ui_auth_stage_complete,
+            session_id,
+            stage_type,
+            result,
+        )
+
+    def _mark_ui_auth_stage_complete(
+        self, txn, session_id: str, stage_type: str, result: Union[str, bool, JsonDict],
+    ):
+        try:
+            # Get the session to ensure it exists.
+            self.db.simple_select_one_txn(
+                txn,
+                table="ui_auth_sessions",
+                keyvalues={"session_id": session_id},
+                retcols=("session_id",),
+            )
```

Comment:
in theory this could race against a deletion? maybe we should `select for update`?

Reply:
When we delete from `ui_auth_sessions` we also delete from `ui_auth_sessions_credentials`, so either: 
1. the auth session has already been deleted and so this will throw, or
2. we have a concurrent delete going on, in which case there will be a lock on the appropriate row in `ui_auth_sessions_credentials` (and so one of the transactions will get blocked or rolled back).

(You could get rid of the select by adding a foreign key to `ui_auth_sessions_credentials` or prefilling the `ui_auth_sessions_credentials` with the appropriate rows so that you only need to do an `UPDATE` and so avoid risking recreating a row)

Reply:
After a bit of back and forth in #synapse-dev:matrix.org I ended up doing the following:

1. Adding a foreign key between the two tables.
2. Removing the `SELECT` clause here (since it is now enforced by the database).

Note that this is still an `UPSERT` to allow for idempotency of API calls (particularly the registration call).